lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -2018,6 +2016,47 @@ void WireOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
 // Statements
 //===----------------------------------------------------------------------===//
 
+/// If the connect is for RefType, implement the constraint for downward only
+/// references. We cannot connect :
+///   1. an input reference port to the output reference port.
+///   2. instance reference results to each other.
+/// This means, the connect can only be used for forwarding RefType module
+/// ports to Instance ports.
+static LogicalResult checkRefTypeFlow(Operation *connect) {
+  Value dst = connect->getOperand(0);
+  Value src = connect->getOperand(1);
+
+  if (dst.getType().isa<RefType>()) {
+    if (getDeclarationKind(src) == getDeclarationKind(dst)) {
+      auto srcRef = getFieldRefFromValue(src);
+      bool rootKnown;
+      auto srcName = getFieldName(srcRef, rootKnown);
+      auto diag = emitError(connect->getLoc())
+                  << "connect has invalid flow for ref type ports: the source "
+                     "expression ";
+      if (rootKnown)
+        diag << "\"" << srcName << "\" ";
+      diag << "and destination expression ";
```

Comment:
```suggestion
      diag << "and ";
```

Reply:
Thanks! The checks around this for `rootKnown` and `dstRootKnown` make it seem possibly this message won't be able to print info about one or both operands (perhaps more likely if there's something wrong happening?), and the message should read reasonably if they're missing.  I'll give it a go, but mostly what you suggested, appreciate it.